server/aesdsocket.c

```diff
@@ -0,0 +1,340 @@
+	
+        // allocate memory for buffers
+	buffer = (char *)malloc(BUFFER_SIZE*sizeof(char));
+	if(buffer == NULL)
+	{
+		printf("Error while doing malloc\n");
+		return -1;
+	}
+
+	transmit_buffer = (char *)malloc(BUFFER_SIZE*sizeof(char));
+        if(transmit_buffer == NULL)
+        {
+                printf("Error while doing malloc\n");
+                return -1;
+        }
+
+	while(1)
+	{
+	        // accept an incomming connection
+                client_fd = accept(socket_fd, (struct sockaddr *)&their_addr, &addr_size); 
+                if(client_fd < 0)
+	        {
+			printf("Error while accpeting incomming connection\n");
+			close(socket_fd);
+                	close(write_fd);
+                	closelog();
+        	        free(buffer);
+	                free(transmit_buffer);
+	                remove(FILE_PATH);
+			return -1;
+	        }
+        
+	        // convert the IP to a string and print it:
+                inet_ntop(AF_INET, get_in_addr((struct sockaddr *)&their_addr), ipstr, sizeof ipstr);
+                syslog(LOG_DEBUG,"Accepted connection from %s", ipstr);
+
+		loc = 0;
+		
+        	// receive data
+		do
+		{
+	       		received_bytes = recv(client_fd, buffer+loc, BUFFER_SIZE, 0); //sizeof(buffer)
+       			if(received_bytes <= 0)
+			{
+				printf("Error while receiving\n");
+			}
+		
+			loc += received_bytes;
+			
+			// increase buffer size dynamically if required
+			count += 1;
+			buffer = (char *)realloc(buffer, count*BUFFER_SIZE*(sizeof(char)));
+		        if(buffer == NULL)
+			{
+				close(socket_fd);
+	                	close(client_fd);
+        	        	close(write_fd);
+                		closelog();
+                		free(buffer);
+                		free(transmit_buffer);
+                		remove(FILE_PATH);
+				printf("Error while doing realloc\n");
+				return -1;
+			}
+				
+
+		}while(strchr(buffer, '\n') == NULL);
+		received_bytes = loc;
+         	buffer[received_bytes] = '\0';
+		total_bytes += received_bytes; //keep track of total bytes 
+	
+		
+		// write to the file
+	        write_bytes = write(write_fd, buffer, received_bytes);
+                if(write_bytes < 0)
+		{
+			printf("Error in writing bytes\n");
+			close(socket_fd);
+                        close(client_fd);
+                        close(write_fd);
+                        closelog();
+                        free(buffer);
+                        free(transmit_buffer);
+                        remove(FILE_PATH);
+			return -1;
+		}
+		
+		transmit_buffer = (char *)realloc(transmit_buffer, total_bytes*sizeof(char));
+                if(transmit_buffer == NULL)
+                {
+			close(socket_fd);
+                        close(client_fd);
+                        close(write_fd);
+                        closelog();
+                        free(buffer);
+                        free(transmit_buffer);
+                        remove(FILE_PATH);
+                        printf("Error while doing realloc\n");
```

Comment:
@midhun9 rather than duplicate the error logic everywhere, consider a design where you could fall through on error paths and handle cleanup in one place for both error and success paths.